### PR TITLE
IK simulator added

### PR DESCRIPTION
**Pull request by erleben**
This is the first version of an inverse kinematics (IK) simulator. 

It currently contains the basic data structures (bones, skeleton and chain) as well as support for

- Evaluating the forward kinematics: By calling the update_skeleton function.
- Getting the Jacobian of the forward kinematics
- Evaluating the IK objective function
-  Getting the gradient and Hessian of the IK objective function

Current limitations:

- All joints are currently ball-and-socket joints. There is support for different Euler encodings. 
- There are no joint limits implemented into the current framework.
- The unit testing could cover more cases.
- There is no actual IK solver implemented. However, current functions should integrate nicely into SciPy solvers.
- A tutorial is missing in the documentation
- Import/export features with fbx/usd file formats would be nice.
- No visualization has been added.

### CODE REVIEW

File: python/rainbow/simulators/inverse_kinematics/api.py [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+                Bh = skeleton.bones[h]
+
+                q_parent = Q.identity()
+                if Bh.has_parent():
+                    q_parent = skeleton.bones[Bh.parent].q_wcs
+
+                # In the 7 lines of code below the ZYZ Euler angles are hardwired into the code.
+                q_alpha = Bh.get_rotation_alpha()
+                q_beta = Bh.get_rotation_beta()
+                q_gamma = Bh.get_rotation_gamma()
```
Comment (Vakzze) [inline python/rainbow/simulators/inverse_kinematics/api.py:316]: Is this unused?

Comment (Vakzze) [reply python/rainbow/simulators/inverse_kinematics/api.py:316]: q_gamma = Bh.get_rotation_gamma()

**Review by Vakzze — APPROVED** [review]
I have written two questions, otherwise, it looks good to me

File: python/rainbow/simulators/inverse_kinematics/api.py [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+    t_parent = V3.zero()
+    q_parent = Q.identity()
+    if bone.has_parent():
+        t_parent = skeleton.bones[bone.parent].t_wcs
+        q_parent = skeleton.bones[bone.parent].q_wcs
+    bone.t_wcs = t_parent + Q.rotate(q_parent, bone.t)
+    bone.q_wcs = Q.prod(q_parent, q_bone)
+
+    for idx in bone.children:
+        __update_bone(skeleton.bones[idx], skeleton)
```
[the remark this answers is not in the document]
Comment (erleben) [reply python/rainbow/simulators/inverse_kinematics/api.py:129]: Character hierarchies usually have very low depth/height. A humanoid is often modeled using some 30-40 bones and path from root to leaf would have 5-7 bones at most. I would be very surprised if Python recursion depth becomes a challenge. We should be safe.

File: python/rainbow/simulators/inverse_kinematics/api.py (hunk 1 above)
Comment (erleben) [reply python/rainbow/simulators/inverse_kinematics/api.py:316]: Yes, but it is left in the code for didactic purposes. An optimized implementation would likely out comment this line.

**Pull request merged**